src/components/MessageList/renderMessages.tsx

```diff
@@ -0,0 +1,107 @@
+import React, { ReactNode } from 'react';
+
+import { MessageProps } from '../Message/types';
+import { StreamMessage } from '../../context/ChannelStateContext';
+import { DefaultStreamChatGenerics } from '../../types/types';
+import { ComponentContextValue, CustomClasses, isDate } from '../../context';
+import { CUSTOM_MESSAGE_TYPE } from '../../constants/messageTypes';
+import { GroupStyle } from './utils';
+import { Message } from '../Message/Message';
+import { UserResponse } from 'stream-chat';
+
+export interface RenderMessagesOptions<
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+> {
+  components: ComponentContextValue<StreamChatGenerics>;
+  lastReceivedId: string | null;
+  messageGroupStyles: Record<string, GroupStyle>;
+  messageProps: Omit<MessageProps<StreamChatGenerics>, MessagePropsToOmit>;
+  messages: Array<StreamMessage<StreamChatGenerics>>;
+  readData: Record<string, Array<UserResponse<StreamChatGenerics>>>;
+  threadList: boolean;
+  customClasses?: CustomClasses;
+}
+
+export type MessageRenderer<
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+> = (options: RenderMessagesOptions<StreamChatGenerics>) => Array<ReactNode>;
+
+type MessagePropsToOmit =
+  | 'channel'
+  | 'groupStyles'
+  | 'initialMessage'
+  | 'lastReceivedId'
+  | 'message'
+  | 'readBy'
+  | 'threadList';
+
+export function defaultRenderMessages<
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+>({
+  components,
+  customClasses,
+  lastReceivedId,
+  messageGroupStyles,
+  messageProps,
+  messages,
+  readData,
+  threadList,
+}: RenderMessagesOptions<StreamChatGenerics>) {
+  const { DateSeparator, HeaderComponent, MessageSystem } = components;
```

Comment:
I am thinking whether we need `components` param as when we provide our own custom rendered, we can import our custom components there. I think that gradually the integrators may start to require add more and more components to the `components` param.

Reply:
Please ignore, now I see this is the whole component context